python-oauth-web-app/atproto_security.py

```diff
@@ -0,0 +1,41 @@
+from urllib.parse import urlparse
+import requests_hardened
+
+
+# this is a crude/partial filter that looks at HTTPS URLs and checks if they seem "safe" for server-side requests (SSRF). This is only a partial mitigation, the actual HTTP client also needs to prevent other attacks and behaviors.
+# this isn't a fully complete or secure implementation
+def is_safe_url(url):
+    parts = urlparse(url)
+    if not (
+        parts.scheme == "https"
+        and parts.hostname is not None
+        and parts.hostname == parts.netloc
+        and parts.username is None
+        and parts.password is None
+        and parts.port is None
+    ):
+        return False
+
+    segments = parts.hostname.split(".")
+    if not (
+        len(segments) >= 2
+        and segments[-1] not in ["local", "arpa", "internal", "localhost"]
+    ):
+        return False
+
+    if segments[0].isdigit():
```

Comment:
Numeric sub domains should be allowed, no ? Only actual IPs are not. Note that IPv6 are in the `[::1]` form.

Reply:
ah, you are right, it is the TLD (final segment) which can't be all-digits (to prevent confusion with IPv4 addresses)

Reply:
fixed